### PR TITLE
feat(snuba): Implement get_group_event_filter (formerly get_group_eve…

### PR DESCRIPTION
**Pull request by bretthoerner**
…nt_ids)

In order to hit Snuba to search tags within an event the tagstore
interface had to change slightly. This isn't a long term solution,
eventually we'll probably want Event search to be its own service
abstraction (like issue search) and all (or almost all) of the search
can be implemented in a single Snuba query.

For now this kicks the can down the road so we can continue comparison
testing.

### CODE REVIEW

File: src/sentry/api/endpoints/group_events.py [hunk 1]
```diff
@@ -92,17 +92,18 @@ def respond(queryset):
...
+            event_filter = tagstore.get_group_event_filter(
                 group.project_id,
                 group.id,
                 environment.id if environment is not None else None,
                 tags,
             )
 
-            if not event_ids:
+            # HACK: Short-circuit filter responses like {'id__in': []}
+            if len(event_filter) == 1 and not event_filter.values()[0]:
```
Comment (bretthoerner) [inline src/sentry/api/endpoints/group_events.py]: Hm, I should probably just return (and check for) `None` instead.

File: src/sentry/tagstore/snuba/backend.py [hunk 2]
```diff
@@ -380,9 +380,31 @@ def get_groups_user_counts(self, project_id, group_ids, environment_id):
...
+        filters = {
+            'project_id': [project_id],
+            'environment': [environment_id],
+            'issue': [group_id],
+        }
+
+        conditions = [[['tags[{}]'.format(k), '=', v] for (k, v) in tags.items()]]
+
+        result = snuba.query(start, end, groupby=['event_id'], conditions=conditions,
+            filter_keys=filters, referrer='tagstore.get_group_event_filter')
```
Comment (tkaemming) [inline src/sentry/tagstore/snuba/backend.py]: Probably need to slap a limit on this bad boy.

File: src/sentry/tagstore/snuba/backend.py [hunk 3]
```diff
@@ -380,9 +380,31 @@ def get_groups_user_counts(self, project_id, group_ids, environment_id):
...
         return defaultdict(int, {k: v for k, v in result.items() if v})
 
-    def get_group_event_ids(self, project_id, group_id, environment_id, tags):
-        # This method is not implemented since the `event.id` column doesn't
-        # exist in Snuba.
+    def get_group_event_filter(self, project_id, group_id, environment_id, tags):
+        start, end = self.get_time_range()
+        filters = {
+            'project_id': [project_id],
+            'environment': [environment_id],
```
Comment (bretthoerner) [inline src/sentry/tagstore/snuba/backend.py]: This actually feels wrong, as tagstore methods should allow for `environment_id = None` to mean "all environments." But it seems to be all over the Snuba tagstore backend, so I'm going to do a follow on PR with fixes/tests.

File: src/sentry/tagstore/snuba/backend.py [hunk 4]
```diff
@@ -391,3 +413,7 @@ def get_group_ids_for_search_filter(
         # exist in Snuba. This logic is implemented in the search backend
         # instead.
         raise NotImplementedError
+
+    def update_group_for_events(self, project_id, event_ids, destination_id):
+        # Group updates are unncessary in Snuba, but we shouldn't throw an error.
+        pass
```
Comment (tkaemming) [inline src/sentry/tagstore/snuba/backend.py]: This seems unrelated to the rest of this change. Was there some reason to include this here rather than making this change separately?

Comment (bretthoerner) [reply src/sentry/tagstore/snuba/backend.py]: Errr nope, those were all just stubs from my scan of what-methods-we-need-to-add. I can move it out.

File: src/sentry/tagstore/v2/backend.py [hunk 5]
```diff
@@ -922,7 +922,7 @@ def get_group_ids_for_search_filter(
 
         # ANY matches should come last since they're the least specific and
         # will provide the largest range of matches
-        tag_lookups = sorted(six.iteritems(tags), key=lambda (k, v): v == ANY)
+        tag_lookups = sorted(six.iteritems(tags), key=lambda k_v: k_v[1] == ANY)
```
Comment (bretthoerner) [inline src/sentry/tagstore/v2/backend.py]: This is some autopep shit. :(

File: src/sentry/tagstore/snuba/backend.py (hunk 3 above)
Comment (alex-hofsteede) [reply src/sentry/tagstore/snuba/backend.py]: Yeah so the current situation with the util is basically.
`'environment': [None]` gets stripped, so that there are no environment filters at all
`'environment': []` Same thing, no environment filters, all envs returned.
`'environment': ['']`: Gets translated into `environment IS NULL` which gives us just the events with the default environment (because we were saving the default environment as Null in snuba)

Comment (bretthoerner) [reply src/sentry/tagstore/snuba/backend.py]: Ahhh OK. I guess my brain is still so deep in the other implementations that it just read like a bug. Makes sense, I guess this would have been very noticeable in comparisons.

Comment (alex-hofsteede) [reply src/sentry/tagstore/snuba/backend.py]: Yeah its a little gross and I hate mixing all those falsy things like None, empty strings, and empty lists, in python. But it also seemed a little neater than having to conditionally add the environment filter `if environment_id is not None` in all these methods.

**Comment by bretthoerner** [comment]
I think I resolved comments but not sure if y'all consider this one actually reviewed or not? Never got that checkmark.

**Review by tkaemming — APPROVED** [review]
